rs/linux_kernel_command_line/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+        let mut firstpos: Option<usize> = None;
+        self.tokenized_arguments = self
+            .tokenized_arguments
+            .iter()
+            .enumerate()
+            .filter(|(pos, arg)| {
+                let res =
+                    *arg != argument && !arg.starts_with((argument.to_owned() + "=").as_str());
+                if !res && firstpos.is_none() {
+                    firstpos.replace(*pos);
+                }
+                res
+            })
+            .map(|(_, x)| x.clone())
+            .collect::<Vec<String>>();
+        firstpos
+    }
+
+    /// Ensure an argument is present in a kernel command line.
+    /// If the argument is in the command line, it is replaced with the supplied
+    /// value in the same position, and all other mentions are removed.
+    /// If the argument is not in the command line, it is appended to the end.
+    pub fn ensure_single_argument(
+        &mut self,
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<(), UnrepresentableValue> {
+        match self.remove_argument(argument) {
+            Some(pos) => {
+                // match pos >= self.tokenized_arguments.len()
+                self.tokenized_arguments
+                    .insert(pos, Self::format_argument(argument, value)?);
+            }
+            None => {
+                self.add_argument(argument, value)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Add an argument to the end of a kernel command line.
+    /// This does not replace previous arguments of the same name.
+    pub fn add_argument(
+        &mut self,
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<(), UnrepresentableValue> {
+        let to_add = Self::format_argument(argument, value)?;
+        self.tokenized_arguments.push(to_add);
+        Ok(())
+    }
+}
+
+impl From<KernelCommandLine> for String {
+    fn from(val: KernelCommandLine) -> Self {
+        val.tokenized_arguments.join(" ")
+    }
+}
+
+impl TryFrom<&str> for KernelCommandLine {
+    type Error = ImproperlyQuotedValue;
+
+    fn try_from(cmdline: &str) -> Result<Self, ImproperlyQuotedValue> {
+        let mut res: Vec<String> = vec![];
+        let mut curr: Vec<char> = vec![];
+        let mut is_quoted = false;
+        for ch in cmdline.chars() {
+            match ch {
+                '"' => {
+                    if is_quoted {
+                        curr.push(ch);
+                        res.push(curr.iter().collect());
+                        curr = vec![];
+                        is_quoted = false;
+                    } else {
+                        curr.push(ch);
+                        is_quoted = true;
+                    }
+                }
+                ' ' => {
+                    if is_quoted {
+                        curr.push(ch);
+                    } else if !curr.is_empty() {
+                        res.push(curr.iter().collect());
```

Comment:
`res.push(std::mem::take(curr))` does exactly `res.push(curr.iter().collect()); curr = String::new()` but much more efficiently

Even if you don't like `std::mem::take`, prefer `curr.clone()` which better expresses intention and can be more efficient (assuming that you take my recommendation above and use a `String` type for curr).

(I know efficiency doesn't matter a lot in this context but if it's easy to do, why not)

(same above on line 139)

Reply:
I'm not exactly sure how any of that mem::take thing works.  Mind telling me more please?

Your suggested change causes this problem:

```
error[E0308]: mismatched types
   --> rs/linux_kernel_command_line/src/lib.rs:151:49
    |
151 |                         res.push(std::mem::take(curr));
    |                                  -------------- ^^^^ expected `&mut String`, found `Vec<char>`
    |                                  |
    |                                  arguments to this function are incorrect
    |
    = note: expected mutable reference `&mut String`
                          found struct `Vec<char>`
note: function defined here
```

Reply:
LMK how to proceed.  I don't plan to change away from iterating over Unicode chars.  I started with a string implementation and it was buggy in the end.

Reply:
Would this be a blocker?

Reply:
See my comment above. The code creates a `Vec<char>` and then converts it into a `String`. `Vec<char>` stores each character as 4 bytes whereas `String` uses utf8 encoding (variable size/char), so the two are incompatible. However, I think it's unnecessary to use  `Vec<char>` in this case as it makes the code more complex and less efficient.

`std::mem::take(&mut curr)` would allow to take the contents of `curr` and move it into `res`, after which `curr` would be an empty `String`. The beauty of `std::mem::take` is that moving the contents of a `String` is extremely efficient since it doesn't require iterating through the characters. In addition, the code gets simpler.